### PR TITLE
Reforked updated code, make test and travis should work now. Unit testing in progress.

### PR DESCRIPTION
**Pull request by ss-jiang**
My partner reforked your repository and I added our proxy handler implementation on your code that has multithreading since I'm not familiar with updating a fork. It should be adding unit tests from now.

### CODE REVIEW

File: handler.cpp [hunk 1]
```diff
@@ -273,4 +273,156 @@ RequestHandler::Status StatusHandler::HandleRequest(const Request& request, Resp
     return RequestHandler::PASS;
 }
 
+//////////////////////////////////////////////////////////////////////////////////////
+///////////////////////////////// Proxy Handler /////////////////////////////////////
+//////////////////////////////////////////////////////////////////////////////////////
+
+RequestHandler::Status ProxyHandler::Init(const std::string& uri_prefix, const NginxConfig& config) {
+
+    for (unsigned int i = 0; i < config.statements_.size(); i++){
```
Comment (ryanlpeterman) [inline handler.cpp:282]: this is fine and will work, but consider using iterators in the future to iterate over a vector. I believe they are more efficient and likely to be correct independent of stl container.

i.e. for (some_iterator = some_vector.begin(); some_iterator != some_vector.end(); some_iterator++)

File: handler.cpp [hunk 2]
```diff
@@ -273,4 +273,156 @@ RequestHandler::Status StatusHandler::HandleRequest(const Request& request, Resp
...
+        if (config.statements_[i]->tokens_[0] == "host" && config.statements_[i]->tokens_.size() == 2){
+            m_host = config.statements_[i]->tokens_[1];
+        }
+        else if (config.statements_[i]->tokens_[0] == "port" && config.statements_[i]->tokens_.size() == 2){
+            std::cout << "port: " << m_port << std::endl; 
+            m_port = config.statements_[i]->tokens_[1];
+        }
+    }
+    std::cout << "host: " << m_host << std::endl;
+    std::cout << "port: " << m_port << std::endl;
```
Comment (ryanlpeterman) [inline handler.cpp]: These can be taken out now that it works correct? If its too verbose then the production server logs/stdout will be polluted with debugging statements

File: test/handler_test.cpp [hunk 3]
```diff
@@ -70,3 +70,18 @@ TEST(StaticHandlerTest, StaticClassTest){
 
 }
 
+TEST(ProxyHandlerTest, ProxyClassTest) {
+  std::unique_ptr<ProxyHandler> handler(new ProxyHandler); 
+
+  handler->SetHost("ucla.edu");
```
Comment (ryanlpeterman) [inline test/handler_test.cpp:76]: By using the setHost public method, the public API changed which goes aganist the spec I believe. However even if that were not the case, by using the public method to initialize/set the host, we are sidestepping testin gthe ProxyHandler Init function. Might as well just use the Init function to set the host right? Knocks two birds with one stone

File: handler.cpp [hunk 4]
```diff
@@ -273,4 +273,156 @@ RequestHandler::Status StatusHandler::HandleRequest(const Request& request, Resp
...
+    // Form the request. We specify the "Connection: close" header so that the
+    // server will close the socket after transmitting the response. This will
+    // allow us to treat all data up until the EOF as the content.
+    boost::asio::streambuf request;
+    std::ostream request_stream(&request);
+    std::string request_str = "GET " + path + " HTTP/1.0\r\n\r\n";
+
+    request_stream << request_str;
+
+    std::cout << "REQUEST: " << request_str << std::endl;
```
Comment (ryanlpeterman) [inline handler.cpp]: This print statement isn't necessary anymore right?

File: handler.cpp [hunk 5]
```diff
@@ -273,4 +273,156 @@ RequestHandler::Status StatusHandler::HandleRequest(const Request& request, Resp
...
+    // Check if there is a "Location" header, signaling redirect
+    std::string redirect_path = "";
+    for (auto &header: response_ptr->GetHeaders()) {
+        if (header.first == "Location") // Get the redirect path
+            redirect_path = header.second;
+    }
+
+    // Handle redirect
+    if ((response_ptr->GetStatus() == 302) && !redirect_path.empty()) {
+        // Find whether rediret path is http or https
```
Comment (adiraju13) [inline handler.cpp:370]: Looks good very clean!

File: response_parser.cpp [hunk 6]
```diff
@@ -0,0 +1,64 @@
+#include "response_parser.h"
+#include <iostream>
+
+std::unique_ptr<Response> ResponseParser::Parse(std::string response) {
+	auto output_response = std::unique_ptr<Response>(new Response());
```
Comment (adiraju13) [inline response_parser.cpp:5]: Might be nice to break this function up, but there is a time constraint so don't worry too much about it

File: test/response_parser_test.cpp [hunk 7]
```diff
@@ -0,0 +1,53 @@
+#include "../response_parser.h"
+#include "gtest/gtest.h"
+
```
Comment (adiraju13) [inline test/response_parser_test.cpp:3]: Nice test!

**Review by adiraju13 — COMMENTED** [review]
Everything looks pretty decent to me! Feel free to merge any time you guys are ready!

File: handler.cpp [hunk 8]
```diff
@@ -273,4 +273,156 @@ RequestHandler::Status StatusHandler::HandleRequest(const Request& request, Resp
...
+    }
+
+    // Read until EOF, writing data to output as we go.
+    while (boost::asio::read(socket, response,
+          boost::asio::transfer_at_least(1), error)){
+            if (error)
+                break;
+        } 
+
+    // Read the rest of the resopnse which will be the response body
```
Comment (ryanlpeterman) [inline handler.cpp]: comment typo ("response")

File: handler.cpp [hunk 9]
```diff
@@ -273,4 +273,156 @@ RequestHandler::Status StatusHandler::HandleRequest(const Request& request, Resp
...
+  }
+  return nullptr;
+}
+
+RequestHandler::Status ProxyHandler::HandleRequest(const Request& request, Response* response) {
+    std::cout << "\nProxyHandler::HandleRequest" << std::endl;
+
+    // Pass in the request uri
+    std::string request_uri = request.uri();
+    std::unique_ptr<Response> returned_ptr = get_response(request_uri, m_host, "80");
```
Comment (ryanlpeterman) [inline handler.cpp]: why not just replace request_uri with request.uri()? request.uri seems pretty self documenting

File: handler.cpp [hunk 10]
```diff
@@ -273,4 +273,156 @@ RequestHandler::Status StatusHandler::HandleRequest(const Request& request, Resp
...
+
+    if (returned_ptr != nullptr) {
+        response->SetStatus(returned_ptr->GetStatus());
+        for (auto &header: returned_ptr->GetHeaders()) {
+            response->AddHeader(header.first, header.second);
+        }
+        response->SetBody(returned_ptr->GetBody());
+    }
+    return RequestHandler::PASS;
```
Comment (ryanlpeterman) [inline handler.cpp:429]: if returned_ptr == nullptr, shouldn't we return something other than RequestHandler::PASS? I might be misunderstanding the code but I believe it should signify failure instead

File: handler.h [hunk 11]
```diff
@@ -149,4 +150,35 @@ class StatusHandler : public RequestHandler {
...
+
+    Status Init(const std::string& uri_prefix,
+                const NginxConfig& config);
+
+    Status HandleRequest(const Request& request,
+                         Response* response);
+
+    std::unique_ptr<Response> get_response(std::string path, std::string host, std::string port);
+
+    size_t get_type(std::string input_string);
```
Comment (ryanlpeterman) [inline handler.h]: maybe my ctrl-f isnt working but I dont see where the implementation of this function is. What is this for?

File: test/response_parser_test.cpp (hunk 7 above)
Comment (ss-jiang) [reply test/response_parser_test.cpp:3]: Placeholder files, haha

File: response_parser.cpp [hunk 12]
```diff
@@ -0,0 +1,64 @@
+#include "response_parser.h"
+#include <iostream>
+
+std::unique_ptr<Response> ResponseParser::Parse(std::string response) {
```
Comment (ryanlpeterman) [inline response_parser.cpp:4]: appreciate the comments, this code is very understandable

**Comment by kevin-xr-wu** [comment]
everything should be done now, tell us if something goes wrong

**Comment by rishub** [comment]
fixed code based on your comments. all the tests still pass, so should be ready to merge

**Comment by ryanlpeterman** [comment]
Hi. if you're grading this, we were told by Shaghayegh, our TA, to include the conversations we had in messenger to count for credit about the code since we were getting lower grades on the code review comments:
selection_079

**Pull request merged**